File: Trinity.Runtime/platform_src/android/app/src/main/java/org/elastos/trinity/runtime/contactnotifier/comm/RemoteNotificationCommand.java
package org.elastos.trinity.runtime.contactnotifier.comm;

import android.util.Log;

import org.elastos.trinity.runtime.contactnotifier.ContactNotifier;
import org.elastos.trinity.runtime.contactnotifier.RemoteNotificationRequest;
import org.json.JSONObject;

public class RemoteNotificationCommand implements CarrierCommand {
    private CarrierHelper helper;
    private String contactCarrierUserID;
    private RemoteNotificationRequest notificationRequest;
    private CarrierHelper.OnCommandExecuted completionListener;

    RemoteNotificationCommand(CarrierHelper helper, String contactCarrierUserID, RemoteNotificationRequest notificationRequest, CarrierHelper.OnCommandExecuted completionListener) {
        this.helper = helper;
        this.contactCarrierUserID = contactCarrierUserID;
        this.notificationRequest = notificationRequest;
        this.completionListener = completionListener;
    }

    @Override
    public void executeCommand() {
        Log.i(ContactNotifier.LOG_TAG, "Executing remote contact notification command");
        try {
            // Package our remote command
            JSONObject request = new JSONObject();
            request.put("command", "remotenotification");
            request.put("source", "contact_notifier_plugin"); // purely informative
            request.put("key", notificationRequest.key);
            request.put("title", notificationRequest.title);

            if (notificationRequest.url != null)
                request.put("url", notificationRequest.url);

            helper.carrierInstance.sendFriendMessage(contactCarrierUserID, request.toString());

            completionListener.onCommandExecuted(true, null);
        }
        catch (Exception e) {
            e.printStackTrace();
            completionListener.onCommandExecuted(false, e.getLocalizedMessage());
        }
    }
}

File: Trinity.Runtime/platform_src/ios/elastOS/Classes/ContactNotifier/comm/RemoteNotificationCommand.swift
public class RemoteNotificationCommand : CarrierCommand {
    private var helper: CarrierHelper
    private var contactCarrierUserID: String
    private var notificationRequest: RemoteNotificationRequest
    private var completionListener: CarrierHelper.onCommandExecuted

    init(helper: CarrierHelper, contactCarrierUserID: String, notificationRequest: RemoteNotificationRequest, completionListener: @escaping CarrierHelper.onCommandExecuted) {
        self.helper = helper
        self.contactCarrierUserID = contactCarrierUserID
        self.notificationRequest = notificationRequest
        self.completionListener = completionListener
    }

    public func executeCommand() {
        Log.i(ContactNotifier.LOG_TAG, "Executing remote contact notification command")
        do {
            // Package our remote command
            var request = Dictionary<String, Any>()
            request["command"] = "remotenotification"
            request["source"] = "contact_notifier_plugin" // purely informative
            request["key"] = notificationRequest.key
            request["title"] = notificationRequest.title

            if (notificationRequest.url != nil) {
                request["url"] = notificationRequest.url!
            }

            helper.carrierInstance.sendFriendMessage(to: contactCarrierUserID, withString: request.toString())

            completionListener(true, nil)
        }
        catch (let error) {
            print(error)
            completionListener(false, error.localizedDescription)
        }
    }
}
